pkg/controller/controller_utils.go

```diff
@@ -1073,19 +1073,19 @@ func AddOrUpdateTaintOnNode(ctx context.Context, c clientset.Interface, nodeName
 	if len(taints) == 0 {
 		return nil
 	}
-	firstTry := true
+
 	return clientretry.RetryOnConflict(UpdateTaintBackoff, func() error {
 		var err error
 		var oldNode *v1.Node
-		// First we try getting node from the API server cache, as it's cheaper. If it fails
-		// we get it from etcd to be sure to have fresh data.
-		option := metav1.GetOptions{}
-		if firstTry {
```

Comment:
is this firstTry not enough to avoid the race?

Reply:
Yes. there will be problems in the following scenarios:
When the node status changes from not ready to ready, the two goroutines of kube-controller-manager that remove the unreachable noschedule taint and noexecute taint update the taint information at the same time, which may cause overwriting.

Reply:
Even if the node-controller watches the changes of the node, if in a scenario with multiple kube-apiservers, if the old data (without the noschedule taint, the real data is that the noschedule taint has been overwritten) is obtained from the kube-apiserver, the result is that when the node status changes to ready, the noschedule taint is still retained.